api/datasets.go

```diff
@@ -197,14 +199,72 @@ func extensionToMimeType(ext string) string {
 	}
 }
 
-func (h *DatasetHandlers) listHandler(w http.ResponseWriter, r *http.Request) {
-	args := lib.ListParamsFromRequest(r)
-	args.OrderBy = "created"
+// snoop reads from an io.ReadCloser and restores it so it can be read again
+func snoop(body *io.ReadCloser) (io.ReadCloser, error) {
+	if body != nil && *body != nil {
+		result, err := ioutil.ReadAll(*body)
+		(*body).Close()
+
+		if err != nil {
+			return nil, err
+		}
+		if len(result) == 0 {
+			return nil, io.EOF
+		}
+
+		*body = ioutil.NopCloser(bytes.NewReader(result))
+		return ioutil.NopCloser(bytes.NewReader(result)), nil
+	}
+	return nil, io.EOF
+}
+
+var decoder = schema.NewDecoder()
+
+// UnmarshalParams deserialzes a lib req params stuct pointer from an HTTP
+// request
+func UnmarshalParams(r *http.Request, p interface{}) error {
+	// TODO(arqu): this should be set on the global decoder
+	// probably means the decoder should live in the API struct or somewhere similar
+	decoder.IgnoreUnknownKeys(true)
+	if r.Method == http.MethodPost || r.Method == http.MethodPut {
+		defer func() {
+			if defSetter, ok := p.(lib.NZDefaultSetter); ok {
+				defSetter.SetNonZeroDefaults()
+			}
+		}()
+
+		if r.Header.Get("Content-Type") == jsonContentType {
+			body, err := snoop(&r.Body)
+			if err != nil && err != io.EOF {
+				return err
+			}
+			// this avoids resolving on empty body requests
+			// and tries to handle it almost like a GET
+			if err != io.EOF {
+				return json.NewDecoder(body).Decode(p)
+			}
+		}
+	}
+
+	if ru, ok := p.(lib.RequestUnmarshaller); ok {
+		return ru.UnmarshalFromRequest(r)
+	}
+
+	if err := r.ParseForm(); err != nil {
+		return err
+	}
+	return decoder.Decode(p, r.Form)
+}
 
-	args.Term = r.FormValue("term")
+func (h *DatasetHandlers) listHandler(w http.ResponseWriter, r *http.Request) {
+	params := &lib.ListParams{}
+	if err := UnmarshalParams(r, params); err != nil {
```

Comment:
Defaults are now handled in the lib.Params through `UnmarshalFromRequest` and `SetNonZeroDefaults`